x-pack/plugins/cases/public/components/all_cases/use_all_cases_state.tsx

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useCallback, useRef, useState, useEffect } from 'react';
+import { useLocation, useHistory } from 'react-router-dom';
+import { isEqual } from 'lodash';
+
+import useLocalStorage from 'react-use/lib/useLocalStorage';
+import { parse, stringify } from 'query-string';
+
+import { DEFAULT_FILTER_OPTIONS, DEFAULT_QUERY_PARAMS } from '../../containers/use_get_cases';
+import { parseUrlQueryParams } from './utils';
+import { LOCAL_STORAGE_KEYS } from '../../../common/constants';
+
+import type {
+  FilterOptions,
+  PartialFilterOptions,
+  LocalStorageQueryParams,
+  QueryParams,
+  PartialQueryParams,
+  ParsedUrlQueryParams,
+} from '../../../common/ui/types';
+import { useCasesContext } from '../cases_context/use_cases_context';
+
+export const getQueryParamsLocalStorageKey = (appId: string) => {
+  const filteringKey = LOCAL_STORAGE_KEYS.casesQueryParams;
+  return `${appId}.${filteringKey}`;
+};
+
+export const getFilterOptionsLocalStorageKey = (appId: string) => {
+  const filteringKey = LOCAL_STORAGE_KEYS.casesFilterOptions;
+  return `${appId}.${filteringKey}`;
+};
+
+const getQueryParams = (
+  params: PartialQueryParams,
+  queryParams: PartialQueryParams,
+  urlParams: PartialQueryParams,
+  localStorageQueryParams?: LocalStorageQueryParams
+): QueryParams => {
+  const result = { ...DEFAULT_QUERY_PARAMS };
+
+  result.perPage =
+    params.perPage ??
+    urlParams.perPage ??
+    localStorageQueryParams?.perPage ??
+    DEFAULT_QUERY_PARAMS.perPage;
+
+  result.sortField =
+    params.sortField ??
+    urlParams.sortField ??
+    localStorageQueryParams?.sortField ??
+    DEFAULT_QUERY_PARAMS.sortField;
+
+  result.sortOrder =
+    params.sortOrder ??
+    urlParams.sortOrder ??
+    localStorageQueryParams?.sortOrder ??
+    DEFAULT_QUERY_PARAMS.sortOrder;
+
+  result.page = params.page ?? urlParams.page ?? DEFAULT_QUERY_PARAMS.page;
+
+  return result;
+};
+
+const getFilterOptions = (
+  filterOptions: FilterOptions,
+  params: FilterOptions,
+  urlParams?: PartialFilterOptions,
+  localStorageFilterOptions?: PartialFilterOptions
+): FilterOptions => {
+  const severity =
+    params?.severity ??
+    urlParams?.severity ??
+    localStorageFilterOptions?.severity ??
+    DEFAULT_FILTER_OPTIONS.severity;
+  const status =
+    params?.status ??
+    urlParams?.status ??
+    localStorageFilterOptions?.status ??
+    DEFAULT_FILTER_OPTIONS.status;
+
+  return {
+    ...filterOptions,
+    ...params,
+    ...(severity && { severity }),
```

Comment:
Will `severity` and `status` always be defined since we leverage `DEFAULT_FILTER_OPTIONS`? I think we just do:

```
{
  ...
  severity,
  status,
}
```